keras_nlp/samplers/sampler.py

```diff
@@ -312,24 +317,39 @@ def sample(
         # The index of the last non-padding token in prompt. Since all sequences
         # are aligned to the right side, the index is the same for all.
         current_index = max_length - num_steps
+        original_padding_mask = tf.cast(tf.identity(mask), dtype=tf.int32)
 
         def one_step(
             current_index,
             prompt,
             mask,
             cache=None,
         ):
+            end_token_seen = (prompt == end_token_id) & (
```

Comment:
Why is any of this needed on the loop body?

You just need the fancier cond, but all the tokens after the end id will get truncated anyway elsewhere, so we can keep generating on "dead sequences" without any worries. In fact it will be both faster and simpler.

Reply:
good question, actually I am not sure. So we have two options:
1) Only feed the alive sequences into the model
2) feed all sequences into the model, and let truncation do the job.

If we assume that the batch size does not really affect the time of a forward pass, then there is no need to go for 1). But my guess is the smaller the batch size, the faster the computation, especially when the model is large, and batch size is big. So the tradeoff is between time saved from feeding a smaller batch into model and time increased by doing `boolean_mask` and `scatter_nd_update`.

Reply:
But XLA will compile everything with static shapes. The batch size is not variable, the computation will 100% the same with or without this.

So I'm pretty sure this would be added complexity for zero performance gain. The computations run in each "loop body" will be static once they are compiled with XLA.

Reply:
so the `boolean_mask` call will be nullified? My understanding is the input/output shape in XLA are frozen, but I can still tune the shape inside the function, very few confidence tho.